virl2_client/models/auth_management.py

```diff
@@ -0,0 +1,392 @@
+#
+# This file is part of VIRL 2
+# Copyright (c) 2019-2023, Cisco Systems, Inc.
+# All rights reserved.
+#
+# Python bindings for the Cisco VIRL 2 Network Simulation Platform
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import annotations
+
+import time
+from enum import Enum
+from typing import Any, Optional
+
+from httpx import Client
+
+from .resource_pools import ResourcePool
+
+
+class AuthMethod(Enum):
+    LOCAL = "local"
+    LDAP = "ldap"
+
+
+_BASE_URL = "system/auth"
+
+
+class AuthManagement:
+    _METHOD_URL = _BASE_URL + "/config"
+
+    def __init__(self, session: Client, auto_sync=True, auto_sync_interval=1.0):
+        self.auto_sync = auto_sync
+        self.auto_sync_interval = auto_sync_interval
+        self._last_sync_time = 0.0
+        self._session = session
+        self._method = AuthMethod.LOCAL
+        self._managers: dict[AuthMethod, Optional[AuthMethodManager]] = {
+            AuthMethod("local"): None,
+            AuthMethod("ldap"): LDAPManager(self),
+        }
+
+    def sync_if_outdated(self) -> None:
+        timestamp = time.time()
+        if (
+            self.auto_sync
+            and timestamp - self._last_sync_time > self.auto_sync_interval
+        ):
+            self.sync()
+
+    def sync(self) -> None:
+        settings: dict = self._session.get(self._METHOD_URL).json()
+        self._method = AuthMethod(settings.pop("method"))
+        self._last_sync_time = time.time()
+        # Since right now we also get the other settings along with the method, we
+        # might as well save a request and store said settings in the relevant manager
+        if (manager := self._managers[self._method]) is not None:
+            manager._settings = settings
+            manager._last_sync_time = time.time()
+
+    @property
+    def method(self) -> AuthMethod:
+        self.sync_if_outdated()
+        return self._method
+
+    @method.setter
+    def method(self, value: AuthMethod | str):
+        value = AuthMethod(value)
+        self._session.put(self._METHOD_URL, json={"method": value.value})
+        self._method = value
+
+    @property
+    def manager(self) -> AuthMethodManager:
+        """
+        The manager for the current auth method.
+        """
+        # TODO: should we allow users to also configure currently inactive auth methods?
```

Comment:
API allows you to only configure currently set method, same apply for UI. I therefore think user should also be only allowed to update currently activated auth_manager.

Reply:
AFAIK, you can configure all LDAP properties and still set the method to "local" in a single request (via API only).  This is imo a bad behavior and will be restricted eventually (when? no idea)

Reply:
Yeah, I realized in my next comment. As you pointed out it should not be the case in the future, so I don't think it will be an issue if we already implement it properly in the PCL.

Reply:
agreed

Reply:
As we already agreed on other things, this should be and probably will be handled by API once the schema is fixed, therefore we can take the lazy approach and "do nothing" here.

Reply:
Ok, if we leave the errors to the API, we still have two identically simple options:
```python
client.auth_management.method = "ldap"
ldap_manager = client.auth_management.manager
client.auth_management.method = "local"
ldap_manager.verify_tls = True  # 200 OK, method is changed to ldap
```
or
```python
client.auth_management.method = "ldap"
ldap_manager = client.auth_management.manager
client.auth_management.method = "local"
ldap_manager.verify_tls = True  # 400 Bad Request
```
Honestly, this doesn't matter much either way, so feel free to merge @PatrikMosko @tmikuska or if you want this changed, just drop a comment back.

Reply:
I think it should pass currently, if you are changing ldap-related properties even if local is set. Currently we can consider it a "feature", later, once the schema is fixed, there wont be any extra properties for local so it should fail then. So I would say neither of your options, just patch the property silently (even if local is set) and do nothing, that is how our API works now.